### PR TITLE
 rgw: es: support username and password for ES 

### PR DESCRIPTION
**Pull request by theanalyst**
For ES endpoints terminated with a username and password, either via xpack or
fronted by another webserver with http basic auth, we now support "username" and
"password" configurable which should be capable of doing HTTP basic
authentication

Fixes: https://tracker.ceph.com/issues/23655

This is built atop of https://github.com/ceph/ceph/pull/21852 and basically first adds support for headers in get http crs, and then adds these headers as a part of all the ES requests

### CODE REVIEW

**Comment by theanalyst** [comment]
ran a suite at http://pulpito.ceph.com/abhi-2018-06-06_14:42:46-rgw-wip-abhi-testing-3-distro-basic-smithi/ where we had some failures due to centos mirrors and re-run at http://pulpito.ceph.com/abhi-2018-06-11_09:09:36-rgw-wip-abhi-testing-3-distro-basic-smithi/

**Comment by yuriw** [comment]
@theanalyst pls rebase

--- pr 22030 --- pulling https://github.com/theanalyst/ceph.git branch wip-es6-auth
remote: Counting objects: 32, done.
remote: Total 32 (delta 26), reused 26 (delta 26), pack-reused 6
Unpacking objects: 100% (32/32), done.
From https://github.com/theanalyst/ceph
 * branch            wip-es6-auth -> FETCH_HEAD
Auto-merging src/rgw/rgw_sync_module_es.cc
CONFLICT (file/directory): There is a directory with name src/dmclock in 67fe08364154ff9de1c7a4185c629270bc69a90b. Adding src/dmclock as src/dmclock~HEAD
Automatic merge failed; fix conflicts and then commit the result.
Traceback (most recent call last):
  File "/home/yuriw/wip_master/src/script/build-integration-branch", line 62, in <module>
    assert not r
AssertionError

**Comment by theanalyst** [comment]
@yuriw updated..

**Comment by theanalyst** [comment]
@yehudasa can you review this once again, this was the same as the fixes for ES versions > 5.0 + support for ES deployments with a username + pw

**Comment by theanalyst** [comment]
@yehudasa ping, this contains the ES6 support + creating a username/pw for setups that have X-pack configured. I closed the earlier approved https://github.com/ceph/ceph/pull/21852, this contains that pr + one commit for username/pw support

File: src/rgw/rgw_sync_module_es.cc [hunk 1]
```diff
@@ -103,6 +103,55 @@ class ItemList {
...
+
+  void decode_json(JSONObj *obj);
+};
+
+bool operator >= (const ESVersion& v1, const ESVersion& v2)
+{
+  if (v1.major_ver == v2.major_ver)
+    return v1.minor_ver >= v2.minor_ver;
+
+  return v1.major_ver > v2.major_ver;
```
Comment (yehudasa) [inline src/rgw/rgw_sync_module_es.cc]: shouldn't it be `>=` here?

Comment (theanalyst) [reply src/rgw/rgw_sync_module_es.cc]: I guess the first equality should catch this, having said that this does look non intuitive, I'll collapse this to a one liner

**Comment by theanalyst** [comment]
@yehudasa updated to use a std::pair as the ESVersion, and just a wrapper struct to parse the nested version json.

File: src/rgw/rgw_sync_module_es.cc [hunk 2]
```diff
@@ -137,21 +113,35 @@ struct ESInfo {
...
+struct es_version_decoder {
+  // simple wrapper structure to wrap the es version nested type
+  std::string s;
+  void decode_json(JSONObj *obj) {
+    JSONDecoder::decode_json("number",s,obj);
+  }
+
+  ESVersion get_version() {
+      int _major,_minor;
+      sscanf(s.c_str(), "%d.%d", &_major, &_minor);
```
Comment (chardan) [inline src/rgw/rgw_sync_module_es.cc]: Is there any possibility of parse error (including empty input) here? Remember to check the result of sscanf(), if you'd like to use scanf() here and there's any possibility of error. Or, consider using std::stoi(), which also has error detection (note that the older C atoi() doesn't give you a way to do this) and supports std::string directly.

**Comment by theanalyst** [comment]
rebased against current master, also have made the es_version_decoder throw a JSONDecoder error in case of a failed scanf, since the error will be caught by the rest coroutine which will ultimately raise an -EINVAL.

**Comment by theanalyst** [comment]
Passes local testing, @yehudasa can you take a look

**Comment by theanalyst** [comment]
@yehudasa  updated against the current master, and tested locally, can you take a look

**Comment by yehudasa** [comment]
@theanalyst still looks good

**Comment by yehudasa** [comment]
@theanalyst feel free to merge if passing tests. We'll want to backport it to mimic.

File: src/rgw/rgw_sync_module_es.cc (hunk 2 above)
Comment (yehudasa) [reply src/rgw/rgw_sync_module_es.cc]: @theanalyst should at least dump error and set defaults to _major and _minor.

**Comment by yehudasa** [comment]
@theanalyst though note comment by @chardan

File: src/rgw/rgw_sync_module_es.cc [hunk 3]
```diff
@@ -564,18 +565,33 @@ class RGWElasticInitConfigCBCR : public RGWCoroutine {
...
+
       yield {
         string path = conf->get_index_path();
+        ldout(sync_env->cct, 5) << "got elastic version=" << es_info.version.to_str() << dendl;
 
         es_index_settings settings(conf->num_replicas, conf->num_shards);
         es_index_mappings mappings;
+        if (es_info.version >= ESVersion(5,0)) {
+          ldout(sync_env->cct, 0) << "elasticsearch: using text type for string index mappings " << dendl;
+          mappings.string_type = ESType::Text;
```
Comment (yehudasa) [inline src/rgw/rgw_sync_module_es.cc]: @theanalyst note migration instructions here:
https://www.elastic.co/blog/strings-are-dead-long-live-strings
`string` converts into either `text` or `keyword`, depending on the `analyzed` state. We also need to make sure the `index` field is being created correctly (as code is written now, it doesn't create it).

Comment (theanalyst) [reply src/rgw/rgw_sync_module_es.cc]: Thanks for the find, I'll modify dump_type to handle these cases

Comment (theanalyst) [reply src/rgw/rgw_sync_module_es.cc]: I'm seeing the same behaviour even without the patch in ES 5.6, we get the double mapping to both text and keyword. I assume name fields should be text types and most of the fields like IDs should be keywords

Comment (theanalyst) [reply src/rgw/rgw_sync_module_es.cc]: So reading more into this it seems even mapping types are already deprecated and going away in the next ES version. https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html, not sure how exactly we're supposed to be mapping types then. For now I'll decide based on "analyzed" field whether we dump a keyword or a text if the type set is not a string already.

**Comment by yehudasa** [comment]
@theanalyst I did some work on this, see here:
https://github.com/ceph/ceph/pull/26106

**Comment by theanalyst** [comment]
taken up in #26106

**Pull request closed without merging**